src/data/equiv/derangements/basic.lean

```diff
@@ -0,0 +1,265 @@
+/-
+Copyright (c) 2021 Henry Swanson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Henry Swanson
+-/
+import data.equiv.basic
+import data.equiv.option
+import group_theory.perm.option
+
+import tactic.equiv_rw
+
+/-!
+# Derangements on types
+
+In this file we define a predicate and its corresponding set:
+
+* `is_derangement f`: whether the permutation `f` has no fixed points.
+
+* `derangements α`: the set of derangements on a type `α`.
+
+Then we define
+
+* some related predicates: e.g.,
+
+  - `only_possible_fixed_point f a`: either `f` has no fixed points, or `a` is the only fixed point;
+
+  - `exactly_one_fixed_point f a`: `f` has exactly one fixed point, which is `a`;
+
+* equivalences involving various subtypes of `perm α` and `derangements α`: e.g.,
+
+  - `derangements_equiv_sigma_opfp`: an equivalence between `derangements (option α)` and the
+    sigma-type `Σ a : α, {f : perm α // only_possible_fixed_point f a}`.
+-/
+
+open equiv
+
+section definitions
+
+/-- A permutation is a derangement if it has no fixed points. -/
+def is_derangement {α : Type*} (f : perm α) : Prop := (∀ x : α, f x ≠ x)
+
+/-- The set of derangements on `α`. -/
+def derangements (α : Type*) : set (perm α) := {f : perm α | is_derangement f}
+
+/-- The permutation `f` has at most one fixed point, `a`. -/
+def only_possible_fixed_point {α : Type*} (f : perm α) (a : α) : Prop := ∀ x : α, f x = x → x = a
+
+/-- The permutation `f` has exactly one fixed point, `a`. -/
+def exactly_one_fixed_point {α : Type*} (f : perm α) (a : α) : Prop := ∀ x : α, f x = x ↔ x = a
```

Comment:
These shouldn't be in the global namespace. Perhaps put them inside `perm`, and use dot notation?

Reply:
sure sure, can do

Reply:
Are these statements easier to make in terms of `perm.support`?

Reply:
Oh, I didn't know that existed! I'll give that a try

Reply:
The first would be `f.support ⊆ {a}` and the second `f.support = {a}`. These seem attractive as they avoid the need for either of these definitions.

Reply:
Ah, this needs finiteness. @pechersky, you had some work on this in review, right?